File: ruby/lib/woothee/browser.rb
# -*- coding: utf-8 -*-

require 'woothee/dataset'
require 'woothee/util'

module Woothee::Browser
  extend Woothee::Util

  def self.challenge_msie(ua, result)
    return false if ua.index("compatible; MSIE").nil? && ua.index("Trident/").nil?

    version = if ua =~ /MSIE ([.0-9]+);/o
                $1
              elsif ua =~ /Trident\/([.0-9]+); rv ([.0-9]+)/o
                $2
              else
                Woothee::VALUE_UNKNOWN
              end
    update_map(result, Woothee::DataSet.get('MSIE'))
    update_version(result, version)
    true
  end

  def self.challenge_safari_chrome(ua, result)
    return false if ua.index('Safari/').nil?

    version = Woothee::VALUE_UNKNOWN

    if ua =~ /(?:Chrome|CrMo|CriOS)\/([.0-9]+)/o
      # Chrome
      version = $1
      update_map(result, Woothee::DataSet.get('Chrome'))
      update_version(result, version)
      return true
    end

    # Safari
    if ua =~ /Version\/([.0-9]+)/o
      version = $1
    end
    update_map(result, Woothee::DataSet.get('Safari'))
    update_version(result, version)
    true
  end

  def self.challenge_firefox(ua, result)
    return false if ua.index('Firefox/').nil?

    version = if ua =~ /Firefox\/([.0-9]+)/o
                $1
              else
                Woothee::VALUE_UNKNOWN
              end
    update_map(result, Woothee::DataSet.get('Firefox'))
    update_version(result, version)
    true
  end

  def self.challenge_opera(ua, result)
    return false if ua.index('Opera').nil?

    version = if ua =~ /Version\/([.0-9]+)/o
                $1
              elsif ua =~ /Opera[\/ ]([.0-9]+)/o
                $1
              else
                Woothee::VALUE_UNKNOWN
              end
    update_map(result, Woothee::DataSet.get('Opera'))
    update_version(result, version)
    true
  end

  def self.challenge_sleipnir(ua, result)
    return false if ua.index('Sleipnir/').nil?

    version = if ua =~ /Sleipnir\/([.0-9]+)/o
                $1
              else
                Woothee::VALUE_UNKNOWN
              end
    update_map(result, Woothee::DataSet.get('Sleipnir'))
    update_version(result, version)

    # Sleipnir's user-agent doesn't contain Windows version, so put 'Windows UNKNOWN Ver'.
    # Sleipnir is IE component browser, so for Windows only.
    win = Woothee::DataSet.get('Win')
    update_category(result, win[Woothee::KEY_CATEGORY])
    update_os(result, win[Woothee::KEY_NAME])

    true
  end
end

File: python/lib/woothee/browser.py
# -*- coding: utf-8 -*-

import re
from . import dataset
from . import util

def challenge_msie(ua, result):
  if 'compatible; MSIE' not in ua and 'Trident/' not in ua:
    return False
  version = dataset.VALUE_UNKNOWN
  msie = re.search('MSIE ([.0-9]+);', ua)
  trident = re.search('Trident/([.0-9]+); rv ([.0-9]+)', ua)
  if msie:
    version = msie.group(1)
  elif trident:
    version = trident.group(2)
  util.update_map(result, dataset.get('MSIE'))
  util.update_version(result, version)
  return True

def challenge_safari_chrome(ua, result):
  if 'Safari/' not in ua:
    return False
  version = dataset.VALUE_UNKNOWN
  obj = re.search('(?:Chrome|CrMo|CriOS)/([.0-9]+)', ua)
  if obj:
    # Chrome
    version = obj.group(1)
    util.update_map(result, dataset.get('Chrome'))
    util.update_version(result, version)
    return True
  
  # Safari
  obj = re.search('Version/([.0-9]+)', ua)
  if obj:
    version = obj.group(1)
  util.update_map(result, dataset.get('Safari'))
  util.update_version(result, version)
  return True

def challenge_firefox(ua, result):
  if 'Firefox/' not in ua:
    return False
  obj = re.search('Firefox/([.0-9]+)', ua)
  version = obj.group(1) if obj else dataset.VALUE_UNKNOWN
  util.update_map(result, dataset.get('Firefox'))
  util.update_version(result, version)
  return True

def challenge_opera(ua, result):
  if 'Opera' not in ua:
    return False
  obj = re.search('Version/([.0-9]+)', ua)
  version = dataset.VALUE_UNKNOWN
  if obj:
    version = obj.group(1)
  else:
    obj = re.search('Opera[/ ]([.0-9]+)', ua)
    if obj:
      version = obj.group(1)
  util.update_map(result, dataset.get('Opera'))
  util.update_version(result, version)
  return True

def challenge_sleipnir(ua, result):
  if 'Sleipnir/' not in ua:
    return False
  obj = re.search('Sleipnir/([.0-9]+)', ua)
  version = obj.group(1) if obj else dataset.VALUE_UNKNOWN
  util.update_map(result, dataset.get('Sleipnir'))
  util.update_version(result, version)
  # Sleipnir's user-agent doesn't contain Windows version, so put 'Windows UNKNOWN Ver'.
  # Sleipnir is IE component browser, so for Windows only.
  win = dataset.get('Win')
  util.update_category(result, win[dataset.KEY_CATEGORY])
  util.update_os(result, win[dataset.KEY_NAME])
  return True
